File: frontend/src/types/borrow.ts
import type { BaseQuery } from "./baseQuery";
import type { Item } from "./item";
import type { User } from "./user";

export type BorrowStatus =
  | "pending"
  | "rejected"
  | "approved"
  | "borrowed"
  | "returned";

export interface Borrow {
  id: number;
  user: User | null;
  reviewedUser: User | null;
  item: Item | null;
  purpose: string;
  quantity: number;
  officerNote: string | null;
  borrowDate: string;
  returnDate: string;
  status: BorrowStatus;
  reviewAt: string | null;
  isOverdue: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface BorrowQuery extends BaseQuery {
  status?: BorrowStatus;
  startDate?: string;
  endDate?: string;
}

export interface BorrowRequest {
  itemId: number;
  quantity: number;
  purpose: string;
  borrowDate: string | null;
  returnDate: string | null;
}

export interface BorrowForUserRequest extends BorrowRequest {
  userId: number;
}

export interface BorrowError {
  itemId: number;
  quantity: number;
  purpose: string;
  borrowDate: string;
  returnDate: string;
}

export interface BorrowApprovalRequest {
  officerNote: string | null;
}

export interface BorrowCard {
  totalPending: number;
  totalApproved: number;
  totalRejected: number;
}

File: backend/internal/model/borrow.go
// Package model
package model

import (
	"time"

	"gorm.io/gorm"
)

type BorrowStatus string

const (
	BorrowStatusPending  BorrowStatus = "pending"
	BorrowStatusRejected BorrowStatus = "rejected"
	BorrowStatusApproved BorrowStatus = "approved"
	BorrowStatusBorrowed BorrowStatus = "borrowed"
	BorrowStatusReturned BorrowStatus = "returned"
	BorrowStatusCanceled BorrowStatus = "canceled"
)

type Borrow struct {
	ID             uint           `json:"id" gorm:"primaryKey;autoIncrement:true"`
	UserID         uint           `json:"userId" gorm:"not null;index"`
	User           User           `json:"user" gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE"`
	ReviewedUserID *uint          `json:"reviewedUserID" gorm:"index"`
	ReviewedUser   *User          `json:"reviewedUser" gorm:"foreignKey:ReviewedUserID;constraint:OnDelete:SET NULL"`
	ItemID         uint           `json:"itemId" gorm:"not null;index"`
	Item           Item           `json:"item" gorm:"constraint:OnDelete:CASCADE"`
	Purpose        string         `json:"purpose" gorm:"not null"`
	Quantity       int            `json:"quantity" gorm:"not null"`
	OfficerNote    *string        `json:"officerNote" gorm:"type:text"`
	BorrowDate     time.Time      `json:"borrowDate" gorm:"not null;type:date"`
	ReturnDate     time.Time      `json:"returnDate" gorm:"type:date"`
	Status         BorrowStatus   `json:"status" gorm:"type:enum('pending', 'rejected', 'approved', 'borrowed', 'returned', 'canceled');default:pending;not null"`
	ReviewAt       *time.Time     `json:"reviewAt" gorm:"type:date"`
	CreatedAt      time.Time      `json:"createdAt"`
	UpdatedAt      time.Time      `json:"updatedAt"`
	DeletedAt      gorm.DeletedAt `json:"deletedAt" gorm:"index"`
}
